filebeat/prospector/prospector_log.go

```diff
@@ -94,24 +94,35 @@ func (l *Log) Run() {
 	if l.config.CleanRemoved {
 		for _, state := range l.Prospector.states.GetStates() {
 			// os.Stat will return an error in case the file does not exist
-			_, err := os.Stat(state.Source)
+			stat, err := os.Stat(state.Source)
 			if err != nil {
-				// Only clean up files where state is Finished
-				if state.Finished {
-					state.TTL = 0
-					err := l.Prospector.updateState(input.NewEvent(state))
-					if err != nil {
-						logp.Err("File cleanup state update error: %s", err)
-					}
-					logp.Debug("prospector", "Remove state for file as file removed: %s", state.Source)
-				} else {
-					logp.Debug("prospector", "State for file not removed because not finished: %s", state.Source)
+				l.removeState(state)
+				logp.Debug("prospector", "Remove state for file as file removed: %s", state.Source)
+			} else {
+				// Check if existing source on disk and state are the same. Remove if not the case.
+				newState := file.NewState(stat, state.Source)
+				if !newState.FileStateOS.IsSame(state.FileStateOS) {
+					l.removeState(state)
+					logp.Debug("prospector", "Remove state for file as file removed or renamed: %s", state.Source)
 				}
 			}
 		}
 	}
 }
 
+func (l *Log) removeState(state file.State) {
+	// Only clean up files where state is Finished
+	if state.Finished {
+		state.TTL = 0
+		err := l.Prospector.updateState(input.NewEvent(state))
```

Comment:
We could return here and check the error in the caller, and also remove the `else`